### ISSUE TITLE
provider/openstack: Add support for FWaaS Router Insertion

### ISSUE THREAD

**Issue by fatmcgav** [hashicorp/terraform#12488]
@jtopjian I notice you've added support for the `fwaasrouterinsertion` FWaaS router extension into Gophercloud [here](https://github.com/gophercloud/gophercloud/issues/219).

Are you working on adding the functionality into Terraform? 

If not, I'm happy to take a run at it...

**Comment by jtopjian** [comment]
Go for it :)

**Comment by fatmcgav** [comment]
@jtopjian Cool, will do... 

I presume something like this would be a sensible approach: 
https://github.com/hashicorp/terraform/blob/master/builtin/providers/openstack/resource_openstack_compute_instance_v2.go#L414-L422

**Comment by jtopjian** [comment]
@fatmcgav Yeah, that's the first method I would try.

**Comment by fatmcgav** [comment]
@jtopjian Damn, there was me thinking this would be nice and easy... :'(

So I'm struggling with the fact that we're nesting the `firewalls.CreateOpts` in [`FirewallCreateOpts`](https://github.com/hashicorp/terraform/blob/master/builtin/providers/openstack/types.go#L140-L143) in `types.go`.

My initial changes can be seen here: https://github.com/fatmcgav/terraform/commit/5bccc786e6286dd970be6d00635c92edbaac187b

The log output is here: https://gist.github.com/fatmcgav/2b4aa2fa221a128aa2aa3f02cfe3c893
As you can see on Line 10, we end up with a nested `CreateOptsBuilder` that I can't seem to work out how to remove :( 

Thoughts?

Cheers
Gav

**Comment by fatmcgav** [comment]
Hah, and just like that I took a second look at the code, and spotted that you've used `opts.CreateOptsBuilder.ToFirewallMap()` in Gophercloud. 

Implemented the same in TF type, and the creation worked :D
https://github.com/fatmcgav/terraform/commit/140b60cf296e3fda01338bb6368dee7769620a06

Now to test updates...

**Comment by fatmcgav** [comment]
@jtopjian OK, so I think I've got it all working... 

In the end, I had to overload the `Firewall` type to add the `RouterIDs` field.
This effectively removes the use of the `routerinsertion` package aswell. 

Latest code is here: https://github.com/hashicorp/terraform/commit/865a2d888c4b293aac3acaefd2ea1de1878ad94f

Comments welcome. 
Appreciate it needs a bit of a tidy up, remove some of the log lines and commented code, but would be good to have a quick review make sure I'm not doing anything too stupid... ;)
